### DISCUSSION TITLE
mypy doesn't handle property setters correctly

### DISCUSSION THREAD

**Discussion by WyattBlue** [python/typing#1857] [General]
Consider this example:

```python
class A:
    def __init__(self, val: int) -> None:
        self.val = 1

    @property
    def my_property(self) -> int:
        return self.val
    
    @my_property.setter
    def my_property(self, value: int | str) -> None:
        self.val = int(value)

a = A(1)
a.my_property = "2"  # I expect this to be allowed
```

but instead I get this error:
```
test.py:14: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
```

**Comment by JelleZijlstra** [comment]
This is python/mypy#3004.

**Reply by WyattBlue** [reply]
If mypy doesn't support getters/setters being different types (fair enough), why isn't this line `    def my_property(self, value: int | str) -> None:` causing an error because right now, it does nothing and misleading.

**Reply by erictraut** [reply]
The [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html) doesn't say anything about asymmetric descriptors or properties, but other type checkers like pyright [handle them fine](https://pyright-play.net/?code=MYGwhgzhAECCBcAoaLoBMCmAzaB9XAlgHYEAu%2BAFBBiFgDTQBuYI80xpAlNALQB80AHIB7IhiSpJ0arQB0zENAC80AIyJkqAAIAHAE7CdGPaQCemlJhwBbU7n2HjZqjSzd%2B7IqQlSUejKQArnpE0q7yLBbQUVq29gZGJqay1KSkxlFW0HEOic4y9EwsgeKepNAAPtKkeu4CImI%2BvgURiiocFAolnBpgynAUqj1gsjkJTqb9AEQATFOIQA). As Jelle noted, this is a long-standing open issue with mypy. Interestingly, mypy does handle [asymmetric descriptors](https://mypy-play.net/?mypy=latest&python=3.12&gist=0da2ddf6a3c29812eae00838496d1786) if they are not properties.
